drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+    /// sending the overrun error for has already completed. This will cause
+    /// the SP to see an unexpected ROT_IRQ and trigger a CSn pulse. Even if
+    /// there is no overrun, the reply from the prior request will cause the
+    /// SP to either see that the current request is complete, (if its data
+    /// fit in only 2 fifos), or to to get a reply while it's still trying to
+    /// send the current request. In either of these cases, the SP will see an
+    /// unexpected ROT_IRQ and perform a CSn pulse.
+    ///
+    /// Note that it is possible that we got an overrun error and an unexpected
+    /// CSn Assert. In this case, the unexpected CSn assert takes behavioral
+    /// precedance. We do not want to send a reply for all the reasons listed
+    /// above.
+    UnexpectedCsnAssert,
+
+    /// The SP sent us more bytes than `BUF_SIZE`. This is a serious error.
+    /// We must inform the SP.
+    SpProtocolCapabilitiesMismatch,
 }
 
-impl IO {
-    /// Wait for chip select to be asserted then service the FIFOs until end of frame.
-    /// Returns false on spurious interrupt.
-    fn pio(&mut self, transmit: bool) {
-        ringbuf_entry!(Trace::Pio(transmit));
-        let tx_end = self.tx_buf.as_slice().len(); // Available bytes and trailing zeros
-        let rx_end = self.rx_buf.as_slice().len(); // All of the available bytes
-        self.txcount = 0;
-        self.rxcount = 0;
-        self.overrun = false;
-        self.underrun = false;
-
-        if !transmit {
-            // Ensure that unused Tx buffer is zero-filled.
-            self.tx_buf.as_mut().fill(0);
+/// The state of the IO state machine when trying to write a response to the SP
+#[derive(Clone, Copy, PartialEq)]
+pub enum WriteState {
+    /// The RoT has asserted ROT_IRQ, and is waiting for CSn to be asserted
+    /// so it can clock out it's response.
+    WaitingToTransmit,
+
+    /// The RoT is writing it's response
+    Writing,
+
+    /// The response has successfully been written out, ROT_IRQ has been
+    /// de-asserted, and CSn has been de-asserted. We can safely go back to
+    /// reading.
+    ResponseWritten,
+
+    /// The SP has pulsed CSn, which means that CSn was asserted and de-
+    /// asserted with no data clocked at all. This means that the response
+    /// from the RoT was unexpected.
+    Flush,
+
+    /// The RoT did not clock out all its response bytes in time.
+    /// The RoT must de-assert ROT_IRQ, wait for CSn to be de-asserted,
+    /// and then go back to waiting for the next request.
+    Underrun,
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let mut io = configure_spi();
+
+    let mut rx_buf = [0u8; BUF_SIZE];
+    let mut tx_buf = [0u8; BUF_SIZE];
+
+    let mut handler = Handler::new();
+
+    loop {
+        let rx_msg = RxMsg::new(&mut rx_buf[..]);
+        let tx_msg = TxMsg::new(&mut tx_buf[..]);
+        if let Some(response) = io.spi_read(rx_msg, tx_msg, &mut handler) {
```

Comment:
I think this pattern requires a perhaps unintentional, implicit assumption that the requests and responses are either roughly of equal sizes during a SPI transaction  or that the tx and rx buffers themselves are large and ringlike. For example, let's say that the SP is pipelining status requests or some other request where the request is significantly smaller than the the response. Each SPI transaction will only clock out part of the response, and eventually it's possible the responder runs out of space to write its response into the tx buffer. it now has to make a choice what do in this case.

The only statefulness here is that a single frame can be sent or received in a single SPI transaction. If the code on either side detects this is not the case, corrective action is taken. This doesn't mean the SP can't restart or timeout and send a new request, but it does mean that the RoT must notice when this happens and make a decision. I don't think we can ever avoid having the RoT make a decision based on number of frames sent, or at least maximum amount of data it can handle, so a simplifying assumption was made to make the protocol request/response only. That was my goal at least.

Reply:
This has been changed in accordance with what @lzrd wrote. The RX buffer is cleared before each message to be received, and the TX buffer is cleared before each call to the handler.